pkg/genlib/generator_with_template.go

```diff
@@ -0,0 +1,529 @@
+type GeneratorWithTemplate struct {
+	emitFuncs []emitF
+}
+
+func generateTemplateFromField(cfg Config, fields Fields) []byte {
+	templateBuffer := bytes.NewBufferString("{")
+	for i, field := range fields {
+		fieldWrap := fieldValueWrapByType(field)
+		if fieldCfg, ok := cfg.GetField(field.Name); ok {
+			if fieldCfg.Value != nil {
+				fieldWrap = ""
+			}
+		}
+
+		fieldTrailer := []byte(",")
+		if i == len(fields)-1 {
+			fieldTrailer = []byte("}")
+		}
+
+		if strings.HasSuffix(field.Name, ".*") || field.Type == FieldTypeObject || field.Type == FieldTypeNested || field.Type == FieldTypeFlattened {
+			// This is a special case.  We are randomly generating keys on the fly
+			// Will set the json field name as "field.Name.N"
+			N := 5
+			for i := 0; i < N; i++ {
+				if string(fieldTrailer) == "}" && i < N-1 {
+					fieldTrailer = []byte(",")
+				}
+
+				fieldNameRoot := replacer.Replace(field.Name)
+				fieldTemplate := fmt.Sprintf(`"%s.%d": %s{{.%s.%d}}%s%s`, fieldNameRoot, i, fieldWrap, fieldNameRoot, i, fieldWrap, fieldTrailer)
+				templateBuffer.WriteString(fieldTemplate)
+			}
+		} else {
+			fieldTemplate := fmt.Sprintf(`"%s": %s{{.%s}}%s%s`, field.Name, fieldWrap, field.Name, fieldWrap, fieldTrailer)
+			templateBuffer.WriteString(fieldTemplate)
+		}
+	}
+
+	return templateBuffer.Bytes()
+}
+
+func NewGeneratorWithTemplate(template []byte, cfg Config, fields Fields) (*GeneratorWithTemplate, error) {
+	if len(template) == 0 {
+		template = generateTemplateFromField(cfg, fields)
+	}
+
+	tokenizer := regexp.MustCompile(`([^{]*)({{\.[^}]+}})*`)
+	allIndexes := tokenizer.FindAllSubmatchIndex(template, -1)
+
+	var fieldPrefixBuffer []byte
+	var fieldPrefixPreviousN int
+
+	orderedField := make([]string, 0, len(allIndexes))
+	templateFieldMap = make(map[string][]byte, len(allIndexes))
+
+	for _, loc := range allIndexes {
+		var fieldName []byte
+		var fieldPrefix []byte
+
+		if loc[4] > -1 && loc[5] > -1 {
+			fieldName = template[loc[4]+3 : loc[5]-2]
+		}
+
+		if loc[2] > -1 && loc[3] > -1 {
+			fieldPrefix = template[loc[2]:loc[3]]
+		}
+
+		if len(fieldName) == 0 {
+			if template[fieldPrefixPreviousN] == byte(123) {
+				fieldPrefixBuffer = append(fieldPrefixBuffer, byte(123))
+			} else {
+				fieldPrefixBuffer = append(fieldPrefixBuffer, fieldPrefix...)
+			}
+		} else {
+			fieldPrefixBuffer = append(fieldPrefixBuffer, fieldPrefix...)
+			templateFieldMap[string(fieldName)] = fieldPrefixBuffer
+			orderedField = append(orderedField, string(fieldName))
+			fieldPrefixBuffer = nil
+		}
+
+		fieldPrefixPreviousN = loc[2]
+	}
+
+	trailingTemplate = fieldPrefixBuffer
+	// Preprocess the fields, generating appropriate emit functions
+	fieldMap := make(map[string]emitF)
+	for _, field := range fields {
+		if err := bindField(cfg, field, fieldMap, true); err != nil {
+			return nil, err
+		}
+	}
+
+	// Roll into slice of emit functions
+	emitFuncs := make([]emitF, 0, len(fieldMap))
+	for _, fieldName := range orderedField {
```

Comment:
`emitFuncs` will be in the order of the fields as they are in the template.
this will allow us to respect the template in `emit()`: without order enforcement the template will be "shuffled"